pal/src/misc/time.cpp

```diff
@@ -382,16 +372,79 @@ GetTickCount64()
     }
 #else
     {
-        struct timeval tv;    
+        struct timeval tv;
         if (gettimeofday(&tv, NULL) == -1)
         {
             ASSERT("gettimeofday() failed; errno is %d (%s)\n", errno, strerror(errno));
             goto EXIT;
         }
         retval = (tv.tv_sec * tccSecondsToMillieSeconds) + (tv.tv_usec / tccMillieSecondsToMicroSeconds);
     }
-#endif // HAVE_CLOCK_MONOTONIC 
-EXIT:    
+#endif // HAVE_CLOCK_MONOTONIC
+EXIT:
     return retval;
 }
 
+#if defined(_X86_) || defined(__AMD64__) || defined(__x86_64__)
+inline ULONGLONG rdtsc()
+{
+    ULONGLONG H, L;
+    __asm volatile ("rdtsc":"=a"(L), "=d"(H));
+#ifdef _X86_
+    return L;
+#else
+    return ((ULONGLONG)H << 32) | L;
+#endif
+}
+
+static double CPUFreq()
+{
+    struct timeval tstart, tend;
+    ULONGLONG start, end;
+
+    struct timezone tzone;
+    memset(&tzone, 0, sizeof(tzone));
+
+    start = rdtsc();
+    gettimeofday(&tstart, &tzone);
+
+    usleep(1000); // 1ms
+
+    end = rdtsc();
+    gettimeofday(&tend, &tzone);
+
+    ULONGLONG usec = ((tend.tv_sec - tstart.tv_sec)*1e6)
+                + (tend.tv_usec - tstart.tv_usec);
+
+    if (!usec) return 0;
+    return (end - start) / usec;
+}
+
+static ULONGLONG cpu_speed = CPUFreq() * 1e3; // 1000 + 1e6 => ns to ms
```

Comment:
I meant the char "+" in comment appeared to be an arithmetic operator and confused me.